language/tools/transaction-replay/src/transaction_debugger_interface.rs

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use libra_state_view::StateView;
+use libra_types::{
+    access_path::AccessPath,
+    account_address::AccountAddress,
+    account_state::AccountState,
+    account_state_blob::AccountStateBlob,
+    transaction::{Transaction, Version},
+};
+use libradb::LibraDB;
+use std::{convert::TryFrom, path::Path, sync::Arc};
+use storage_interface::DbReader;
+
+pub trait StorageDebuggerInterface {
+    fn get_account_state_by_version(
+        &self,
+        account: AccountAddress,
+        version: Version,
+    ) -> Result<Option<AccountStateBlob>>;
+    fn get_committed_transactions(&self, start: Version, limit: u64) -> Result<Vec<Transaction>>;
+    fn get_latest_version(&self) -> Result<Version>;
+    fn get_version_by_account_sequence(
+        &self,
+        account: AccountAddress,
+        seq: u64,
+    ) -> Result<Option<Version>>;
+}
+
+pub struct LocalDBDebugger(Arc<dyn DbReader>);
+
+impl LocalDBDebugger {
+    pub fn open<P: AsRef<Path> + Clone>(db_root_path: P) -> Result<Self> {
+        Ok(Self(Arc::new(LibraDB::open(db_root_path, true, None)?)))
+    }
+}
+
+pub struct DebuggerStateView<'a, S: StorageDebuggerInterface> {
+    db: &'a S,
+    version: Version,
+}
+
+impl<'a, S: StorageDebuggerInterface> DebuggerStateView<'a, S> {
+    pub fn new(db: &'a S, version: Version) -> Self {
+        Self { db, version }
+    }
+}
+
+impl<'a, S: StorageDebuggerInterface> StateView for DebuggerStateView<'a, S> {
+    fn get(&self, access_path: &AccessPath) -> Result<Option<Vec<u8>>> {
+        if self.version == 0 {
+            return Ok(None);
+        }
+        Ok(
+            match self
+                .db
+                .get_account_state_by_version(access_path.address, self.version)?
+            {
+                Some(blob) => AccountState::try_from(&blob)?
+                    .get(&access_path.path)
+                    .cloned(),
+                None => None,
+            },
+        )
+    }
+
+    fn multi_get(&self, _access_paths: &[AccessPath]) -> Result<Vec<Option<Vec<u8>>>> {
+        bail!("unimplemeneted")
+    }
+
+    fn is_genesis(&self) -> bool {
+        false
+    }
+}
+
+impl StorageDebuggerInterface for LocalDBDebugger {
+    fn get_account_state_by_version(
+        &self,
+        account: AccountAddress,
+        version: Version,
+    ) -> Result<Option<AccountStateBlob>> {
+        Ok(self
+            .0
+            .get_account_state_with_proof_by_version(account, version)?
```

Comment:
I assume this will fail if the local DB doesn't contain the state at `version` Could we create the state by replaying transactions up to `version` instead of failing here?

Reply:
I'm not sure how this restore part is going to work. IIRC the LibraDB will do the trim every month so this tool should be able to allow us to look into transactions that happened in the past month. @msmouse do you have any clues on how this could happen in production?

Reply:
> I assume this will fail if the local DB doesn't contain the state at version Could we create the state by replaying transactions up to version instead of failing here?

We can't replay backwards, so when a state is pruned, we need to use use a backup and restore to the specific version to an empty folder.

> IIRC the LibraDB will do the trim every month

no, it trims continuously according to this config (keeping 40 Million versions by default, that's about 350GB of disc space): 
https://github.com/libra/libra/blob/650983ad5bb3098b75d2aede6ad47ca6c977cc68/config/src/config/storage_config.rs#L34-L36

Reply:
Does that mean that json-rpc can only respond to queries within that time window?

Reply:
Yes, otherwise error.